src/p4p/test/system_client_server.py

```diff
@@ -0,0 +1,631 @@
+        self.assertEqual(len(meta.fields), 4)
+        fields_code = dict(meta.type().aspy()[2])["fields"]
+        self.assertEqual(fields_code, "as")
+
+    # Equivalent to:
+    #   pvget TESTCOUNTER.counter -r ""
+    def testGetDottedNTScalar(self):
+        # Use the one with unwrapping on
+        ctxt = Context("pva")
+        self.addCleanup(ctxt.close)
+        counter = ctxt.get("TESTCOUNTER.counter")
+        self.assertIsInstance(counter, ntfloat)
+        self.assertEqual(counter, 0.0)
+        self.assertEqual(counter.severity, 0)
+        self.assertEqual(counter.raw.getID(), "epics:nt/NTScalar:1.0")
+
+    # Equivalent to:
+    #   pvget TESTCOUNTER.counter -r ""
+    def testGetDottedNTScalarNoUnwrap(self):
+        counter = self.ctxt.get("TESTCOUNTER.counter")
+        self.assertEqual(counter.getID(), "epics:nt/NTScalar:1.0")
+        self.assertEqual(counter.value, 0.0)
+        self.assertEqual(counter.timeStamp.getID(), "time_t")
+        self.assertEqual(counter.alarm.getID(), "alarm_t")
+        self.assertEqual(counter.alarm.severity, 0)
+
+    def assertCounter(self, value):
+        if self.MALCOLM:
+            counter = self.ctxt.get("TESTCOUNTER.counter").value
+        else:
+            counter = self.counter_structure.counter.value
+        self.assertEqual(counter, value)
+
+    # Should be equivalent to (except pvput doesn't do the right thing):
+    #   pvput TESTCOUNTER -r "counter.value" 5
+    def testPut(self):
+        self.assertCounter(0)
+        self.ctxt.put("TESTCOUNTER", 5, "counter.value")
+        self.assertCounter(5)
+        self.ctxt.put("TESTCOUNTER", 0, "counter.value")
+        self.assertCounter(0)
+
+    # Equivalent to:
+    #   pvput TESTCOUNTER.counter 5
+    def testPutDotted(self):
+        self.assertCounter(0)
+        # TODO: pvput defaults the request to "value", should p4p do the same?
+        self.ctxt.put("TESTCOUNTER.counter", 5, "value")
+        self.assertCounter(5)
+        self.ctxt.put("TESTCOUNTER.counter", 0, "value")
+        self.assertCounter(0)
+
+    # Equivalent to:
+    #   pvget -m TESTCOUNTER -r ""
+    def testMonitorEverythingInitial(self):
+        q = Queue()
+        m = self.ctxt.monitor("TESTCOUNTER", q.put)
+        self.addCleanup(m.close)
+        counter = q.get()
+        self.assertStructureWithoutTsEqual(str(counter), str(counter_expected))
+        self.assertTrue(counter.asSet().issuperset({
+            "meta", "meta.fields", "counter", "zero"}))
+        self.ctxt.put("TESTCOUNTER.counter", 5, "value")
+        counter = q.get()
+        self.assertEqual(counter.counter.value, 5)
+        if self.MALCOLM:
+            # bitsets in pvaPy don't work, so it is everything at the moment
+            self.assertTrue(counter.asSet().issuperset({
+                "meta", "meta.fields", "counter", "zero"}))
+        else:
+            # TODO: timeStamp userTag didn't change value, but timeStamp did,
+            # should it be in the bitset?
```

Comment:
Firstly, the ```Value.changed()``` method will give the right results here (True) as it checks the leaf and parent field changed bits.

I can certainly see the confusion wrt. ```Value.asSet()```.  Right now it uses a direct translation of bit index to field name (intended for debugging).  As this method name should probably change, now may be a good time to re-define it.  So the question is how it should work.

What did you expect to see here?

Reply:
I'm not sure. I set timeStamp.secondsPastEpoch to something new, then set userTag to the same value it was before. I guess this means both should be in the bitset of changes...

I think the current behaviour is probably right.

Reply:
Ok, then I'll just do a rename.  Something more descriptive like ```Value.changedSet()```.